src/storage/src/render/sources.rs

```diff
@@ -199,247 +198,228 @@ where
+                            scope.clone(),
+                            &base_source_config,
+                            rehydrated_token,
+                            refine_antichain(&resume_upper),
+                            &probe,
+                        );
+                    } else {
+                        drop(rehydrated_token)
+                    };
+
+                    // If backpressure is enabled, we probe the upsert operator's
+                    // output, which is the easiest way to extract frontier information.
+                    let upsert = match feedback_handle {
+                        Some(feedback_handle) => {
+                            probe.connect_loop(feedback_handle);
+                            upsert.as_collection()
+                        }
+                        None => upsert.as_collection(),
+                    };
 
                     (
-                        upsert_ok.as_collection(),
-                        Some(upsert_err.as_collection()),
-                        health_update,
+                        upsert.leave(),
+                        health_update
+                            .map(|(index, update)| HealthStatusMessage {
+                                index,
+                                namespace: StatusNamespace::Upsert,
+                                update,
+                            })
+                            .leave(),
                     )
-                }
-                SourceEnvelope::None(none_envelope) => {
-                    let results = append_metadata_to_value(decoded_stream);
-
-                    let flattened_stream = flatten_results_prepend_keys(none_envelope, results);
-
-                    let (stream, errors) = flattened_stream.inner.ok_err(split_ok_err);
+                },
+            );
 
-                    let errors = errors.as_collection();
-                    (stream.as_collection(), Some(errors), empty(scope))
-                }
-                SourceEnvelope::CdcV2 => unreachable!(),
-            };
+            let (upsert_ok, upsert_err) = upsert.inner.ok_err(split_ok_err);
 
             (
-                envelope_ok,
-                envelope_err,
-                decode_health.concat(&envelope_health),
+                upsert_ok.as_collection(),
+                Some(upsert_err.as_collection()),
+                health_update,
             )
         }
+        SourceEnvelope::None(none_envelope) => {
+            let results = append_metadata_to_value(decoded_stream);
+
+            let flattened_stream = flatten_results_prepend_keys(none_envelope, results);
+
+            let (stream, errors) = flattened_stream.inner.ok_err(split_ok_err);
+
+            let errors = errors.as_collection();
+            (stream.as_collection(), Some(errors), empty(scope))
+        }
+        SourceEnvelope::CdcV2 => {
+            let (oks, token) = render_decode_cdcv2(&decoded_stream);
```

Comment:
You mention that this can work with any decoded stream whose schema is of the expected shape––I'm assuming this is part of the `key_desc`/`value_desc` work on the source trait? i.e. we will enforce that but are probably not using that mechanism explicitly right now. Is that right or am I missing something?

Reply:
That is correct, the type checking for CDCv2 envelopes will be in the next PR